### PR TITLE
[Functional unit expansion] Adding floating-point adder and fixed-point MAC-ALU units.

### PR DESCRIPTION
**Pull request by rp15**
All source files, wrappers, and unit tests are included in this pull request. The top-level CGRA test in cgra/test/CGRA_custom_test.py tests a 2x2 homogeneous CGRA that includes the fixed-point ALU-MAC in every tile.

### CODE REVIEW

File: fu/ALUgen_Ron_2nd/ALUgenMAC2nd.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    # Components
+    s.fALU = ALUgenMAC()
+    s.fALU.rhs_2 //= lambda:   0 if s.recv_opt.msg.ctrl == OPT_ADD     else \
+                             ( 1 if s.recv_opt.msg.ctrl == OPT_SUB     else \
+                             ( 3 if s.recv_opt.msg.ctrl == OPT_LT      else \
+                             ( 7 if s.recv_opt.msg.ctrl == OPT_GTE     else \
+                             (10 if s.recv_opt.msg.ctrl == OPT_GT      else \
+                             (14 if s.recv_opt.msg.ctrl == OPT_LTE     else \
+                             (16 if s.recv_opt.msg.ctrl == OPT_MUL     else \
+                             (48 if s.recv_opt.msg.ctrl == OPT_MUL_ADD else 0)))))))
```
Comment (tancheng) [inline fu/ALUgen_Ron_2nd/ALUgenMAC2nd.py:40]: Would the generated logic cause bad timing/long critical path? Do you think use if/else inside the update block would help? Or a smarter mux should be modeled and used?

File: lib/opt_type.py
Comment (tancheng) [inline lib/opt_type.py:1]: Can you also remove the OPT_LE?

File: fu/ALUgen_Ron_2nd/ALUgenMAC2nd.py
Comment (tancheng) [inline fu/ALUgen_Ron_2nd/ALUgenMAC2nd.py:1]: What does 2nd mean here?

File: cgra/test/CGRARon_FP_test.py
Comment (tancheng) [inline cgra/test/CGRARon_FP_test.py:1]: If there is still some corner cases failing, can you add a TODO or highlight them in the corresponding test files?

**Review by tancheng — COMMENTED** [review]
Added some comments, please fix.

File: cgra/test/CGRARon_FP_test.py
[the remark this answers is not in the document]
Comment (rp15) [reply cgra/test/CGRARon_FP_test.py:1]: Renamed to CGRA_custom_test.py.

**Comment by tancheng** [comment]
Also, please make sure the integration test would pass. Now it shows `All checks have failed`. You can look into the checks, it shows multiple errors, e.g.,
```
E   FileNotFoundError: [Errno 2] No such file or directory: '/home/runner/work/VectorCGRA/VectorCGRA/fu/pymtl3_Ron_fixed/svsrc/DW02_multp.v'
```
and
```
E   Expected : 469a.0.0.0
E   Received : 4299.0.0.0
```

File: fu/ALUgen_Ron_2nd/ALUgenMAC2nd.py (hunk 1 above)
Comment (rp15) [reply fu/ALUgen_Ron_2nd/ALUgenMAC2nd.py]: As you suggested, we can optimize this later. I would use an and/or MUX here since the select line is one-hot. I will need to ramp up on how basic RTL constructs are implemented in PyMTL.

Comment (rp15) [reply fu/ALUgen_Ron_2nd/ALUgenMAC2nd.py]: assign s.fALU.rhs_2 = (6'd0 & {6{s.recv_opt.msg.ctrl == OPT_ADD}} |
                                      (6'd1 & {6{s.recv_opt.msg.ctrl == OPT_SUB}} | ...

File: cgra/test/CGRARon_FP_test.py
Comment (rp15) [reply cgra/test/CGRARon_FP_test.py:1]: Removed failing units for now.

File: fu/ALUgen_Ron_2nd/ALUgenMAC2nd.py
Comment (rp15) [reply fu/ALUgen_Ron_2nd/ALUgenMAC2nd.py:1]: Renamed.

File: fu/ALUgen_Ron_2nd/ALUgenMAC2nd.py (hunk 1 above)
Comment (tancheng) [reply fu/ALUgen_Ron_2nd/ALUgenMAC2nd.py]: nit: i suggest we don't use hardcoded `6` in the RTL module, as it would make it difficult to parameterize or later if we want to increase the number of opcodes.
